tensorflow-keras/src/main/java/org/tensorflow/keras/optimizers/AdaDelta.java

```diff
@@ -0,0 +1,200 @@
+    super(assertGraph(tf), learningRate, rho, epsilon);
+    initConfig(learningRate, rho, epsilon);
+  }
+
+  /**
+   * Create an Adadelta optimizer
+   *
+   * @param tf the tensorflow Ops
+   * @param name the name of the Optimizer, defaults to "Adadelta"
+   * @param learningRate The learning rate
+   * @param rho The decay rate.
+   * @param epsilon A constant epsilon used to better conditioning the grad update.
+   */
+  public AdaDelta(Ops tf, String name, float learningRate, float rho, float epsilon) {
+    super(assertGraph(tf), name, learningRate, rho, epsilon);
+    initConfig(learningRate, rho, epsilon);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  protected Optional<Op> prepare(String name) {
+    switch (initializers.size()) {
+      case 0:
+        return Optional.empty();
+      case 1:
+        return Optional.of(initializers.get(0));
+      default:
+        return Optional.of( tf.withSubScope(name).withControlDependencies(initializers).noOp());
+    }
+  }
+
+  /**
+   * Create an Adam Optimizer from a config object
+   *
+   * @param tf the tensorflow Ops
+   * @param config a config object to initialize, he config object has keys for "name",
+   *     "learning_rate", "rho" and "epsilon". If a key is missing the default value is used.
+   */
+  public static AdaDelta fromConfig(Ops tf, Map<String, Object> config) {
+    return create(tf, config);
+  }
+
+  /**
+   * Create an Adadelta optimizer
+   *
+   * @param tf the tensorflow Ops
+   * @param config a config object to initialize, the config
+   *     object has keys for "name", "learning_rate", "rho" and "epsilon". If a key is missing the
+   *     default value is used.
+   */
+  public static AdaDelta create(Ops tf, Map<String, Object> config) {
+    String name = (String) config.get(NAME_KEY);
+    float learningRate = (float) config.getOrDefault(LEARNING_RATE_KEY, LEARNING_RATE_DEFAULT);
+    float rho = (float) config.getOrDefault(RHO_RATE_KEY, RHO_DEFAULT);
+    float epsilon = (float) config.getOrDefault(EPSILON_KEY, EPSILON_DEFAULT);
+    if (name == null) // doe this to get the default name
+    {
+      return new AdaDelta(tf, learningRate, rho, epsilon);
+    } else {
+      return new AdaDelta(tf, name, learningRate, rho, epsilon);
+    }
+  }
+
+  /**
+   * Initialize the configuration based on which constructor is called.
+   *
+   * @param learningRate The learning rate
+   * @param rho The decay rate.
+   * @param epsilon A constant epsilon used to better conditioning the grad update.
+   */
+  private void initConfig(float learningRate, float rho, float epsilon) {
+    this.learningRate = learningRate;
+    config.put(NAME_KEY, this.getOptimizerName());
+    config.put(LEARNING_RATE_KEY, learningRate);
+    config.put(RHO_RATE_KEY, rho);
+    config.put(EPSILON_KEY, epsilon);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Map<String, Object> getConfig() {
+    return config;
```

Comment:
If we have a concrete type for our config, then either our Keras optimizers classes should be parametized to return the right type of config, or either it should be returned as an `Object` itself. I can't see in this PR what is the use of `getConfig()` so I can't give a clear answer.

Reply:
See my previous comment on Config.

Reply:
I have removed `config` from this based on deferring the question to when we address serializing the Keras model.